### ISSUE TITLE
Binding on input [type] breaks number

### ISSUE THREAD

**Issue by GuyShaanan** [angular/angular#13243]
**I'm submitting a ...** 
```
[x] bug report => search github for a similar issue or PR before submitting
```

**Current behavior**
Try to bind the `type` attribute of input:

```js
// in Class
inputType="number"

// in html
<input [type]="inputType" formControlName="integerB" />
```
Initially, the value is presented as number when doing `form.value`:

```js
{"integerB": 5}
```
Now, try to change the input's value.

**Its type has changed to string:**

```js
{"integerB": "5"}
```

while it seems the html attributes haven't changed:
```html
// chrome dev console
<input formcontrolname="integerB" ng-reflect-name="integerB" ng-reflect-type="number" type="number" class="ng-valid ng-dirty ng-touched">
```

**Expected behavior**
If we *are* able to bind to [type], the input type should have stayed "number", so does the value.

**Minimal reproduction of the problem with instructions**
http://plnkr.co/edit/kp0xoDbZmJnD5orOTZUj?p=preview

**What is the motivation / use case for changing the behavior?**
I'm dynamiclly generating forms from given JSON, so binding to [type] can save a lot of copy-paste code.

**Please tell us about your environment:**
Not sure this is relevant as it is reproducible in plnkr.

Thanks.

**Comment by DzmitryShylovich** [comment]
Duplicate of https://github.com/angular/angular/issues/7329

**Comment by GuyShaanan** [comment]
Looks like a dup indeed, although my issue is not restricted to ngModel only but also for ReactiveForms.

The [Dynamic Forms](https://angular.io/docs/ts/latest/cookbook/dynamic-form.html) demo still demonstrates this behaviour (the ability to bind to [type]) and is misleading.

**Comment by DzmitryShylovich** [comment]
I would say it's a compiler issue. Angular instantiates directive before resolve inputs so instead of NamberValueAccessor it initializes DefaultValueAccessor. And we can't do much with it inside forms module.

> The Dynamic Forms demo still demonstrates this behaviour (the ability to bind to [type]) and is misleading.

they should remove it from the docs.

**Comment by DzmitryShylovich** [comment]
<input formControlName="integerB" [type]="'number'"/> 
<input formControlName="integerB" [attr.type]="'number'"/>
doesn't trigger creation of NumberValueAccessor https://plnkr.co/edit/u9RAULnbb4lGTR6JJhos?p=preview So it's blocked by https://github.com/angular/angular/issues/11716

**Comment by michael-letcher** [comment]
Is this being fixed any time soon? It's a real gapping hole in a "big" feature of angular.

Check #21418 for a workaround I use.

**Comment by mounikaymrs** [comment]
Wonder why this still hasn't been fixed. It seems like a common use case and see many requests to have this support.

**Comment by michael-letcher** [comment]
It's very surprising, I guess not many use reactive forms or were smart enough not to use Angular 5.

**Comment by ceor** [comment]
Is there another Angular way of doing forms?

**Comment by gorkemyontem** [comment]
Wow, I am also wondering why this still hasn't been fixed. This bug makes it really hard to create a dynamic form.

**Comment by michael-letcher** [comment]
@gorkemyontem You have no idea, our dynamic form field component is huge so that we can cover each broken outlier.

**Comment by michael-letcher** [comment]
@ceor there is a workaround referenced here #21418

**Comment by michaeljota** [comment]
I have done several dynamic input fields, and I have learn that each type of input field is different, and even when you seem to only check for the input type, the fact that the browser renders a different element for each makes you appreciate this kind of behaviors.

**Comment by gitbnw** [comment]
The deprecation warning you get with ng-model makes the workaround really confusing.

**Comment by belyan** [comment]
> @ceor there is a workaround referenced here #21418, still works in Angular 6

Could you please provide an example of code with WA?

**Comment by dinohorvat** [comment]
I've been searching for days for this. Are there any plans to fix this?
Here's another example:

https://stackblitz.com/edit/angular-rzttpk?file=src/app/app.component.ts

**Comment by tremppu** [comment]
Still an issue with Angular 9...

## 🔬 Minimal Reproduction
https://stackblitz.com/edit/angular-number-input-bug

**Comment by PedroFerr** [comment]
No wonder... why the entire World is changing for React.

**Comment by michael-letcher** [comment]
@PedroFerr It is rather sad.

**Comment by yoh-m** [comment]
Still an issue? Yup, still an issue.

**Comment by paul61843** [comment]
This problem has been exited for fove years

**Comment by ekabelly** [comment]
> No wonder... why the entire World is changing for React.

every framework/libarary has bugs

**Comment by HackPoint** [comment]
> No wonder... why the entire World is changing for React.

@PedroFerr it's because they don't that react was built on angular 😉

**Comment by inorganik** [comment]
Giving this a bump because I think it's still an important issue to fix.

**Comment by bllctnr** [comment]
I just started coding with angular framework, It's surprising that this issue is still unresolved.

**Comment by mtuzinskiy** [comment]
Not fixed since 2016?  Seriously?

**Comment by Anutrix** [comment]
It's 2022 already. Is this gonna be noticed anytime this year?

**Comment by Almostsagar** [comment]
Not fixed???? 2022

**Comment by Almostsagar** [comment]
Not fixed?

**Comment by Ghandie** [comment]
April 2022, still no fix in sight?

**Comment by DFG-kjtroll** [comment]
Until this is fixed, you can set a [ngSwitch] on a parent element to the variable holding the input type, and use multiple input tags each with *ngSwitchCase. Not as clean, but it works.

```html
<ng-container [ngSwitchCase]="inputTypeVariable">
  <input *ngSwitchCase="'text'" type="text" ...>
  <input *ngSwitchCase="'number'" type="number" ...>
</ng-container>
```

**Comment by royeradames** [comment]
> ```
> *ngSwitchCase
> ```

Sorry to report that it doesn't work

**Comment by royeradames** [comment]
> Until this is fixed, you can set a [ngSwitch] on a parent element to the variable holding the input type, and use multiple input tags each with *ngSwitchCase. Not as clean, but it works.
> 
> ```
> <ng-container [ngSwitch]="inputTypeVariable">
>   <input *ngSwitchCase="'text'" type="text" ...>
>   <input *ngSwitchCase="'number'" type="number" ...>
>   ...
> </ng-container>
> ```

It didn't work for me

**Comment by DFG-kjtroll** [comment]
> It didn't work for me. I will have to make a duplicate that only differ by the input

It does work, you must be doing something else wrong. You have to put multiple input tags within the [ngSwitch] container, one for each possible case, explicitly declare the type in each tag, and repeat everything. Which is why it's just a sloppy workaround, if this type binding issue gets fixed you could accomplish it with just one input tag.

So to dynamically generate multiple form fields, you could have something like this in the component model:
```ts
private formFields: {
  inputType: 'text' | 'number' | 'date'...
  value: any
}[]
```
Then something like this in the template:
```html
<ng-container *ngFor="let field of formFields"
  <ng-container [ngSwitch]="field.inputType">
    <input *ngSwitchCase="'text'" type="text"  [value]="field.value"...>
    <input *ngSwitchCase="'number'" type="number" [value]="field.value"...>
  ...
  </ng-container>
</ng-container>
```

**Comment by maaz-siddiqui-azakaw** [comment]
7 years and counting

**Comment by H-dymitr** [comment]
Hello 2023

**Comment by Saul-BT** [comment]
React ftw

**Comment by FilipeFerracini** [comment]
> Until this is fixed, you can set a [ngSwitch] on a parent element to the variable holding the input type, and use multiple input tags each with *ngSwitchCase. Not as clean, but it works.
> 
> ```
> <ng-container [ngSwitch]="inputTypeVariable">
>   <input *ngSwitchCase="'text'" type="text" ...>
>   <input *ngSwitchCase="'number'" type="number" ...>
>   ...
> </ng-container>
> ```

This is the way 😢

**Comment by yuriybelike** [comment]
Amazing how this "enterprise" framework can't get forms right, even after 17 versions (some of them with breaking changes, so there is no backward compatibility excuse).

In what sane world is

```html
<input [type]="'number'">
```

different from

```html
<input type="number">
```
?

**Comment by DavidLeonard77** [comment]
You have got to be joking! This is what the workaround looks like in my code..

```
<ng-container [ngSwitch]="inputTypeVariable">
  <input
    *ngSwitchCase="'text'"
    type="text"
    #inputRef
    class="input-field typography-body-input"
    [formControl]="formControl"
    [id]="controlName"
    [autocomplete]="autocomplete"
    [placeholder]="placeholder"
    [readonly]="readonly"
    [attr.aria-required]="required"
    [attr.aria-invalid]="formControl.invalid"
    [attr.aria-disabled]="formControl.disabled"
    (focus)="inputFocus($event)"
    (blur)="inputBlur($event)"
  />
  <input
    *ngSwitchCase="'number'"
    type="number"
    #inputRef
    class="input-field typography-body-input"
    [formControl]="formControl"
    [id]="controlName"
    [autocomplete]="autocomplete"
    [placeholder]="placeholder"
    [readonly]="readonly"
    [attr.aria-required]="required"
    [attr.aria-invalid]="formControl.invalid"
    [attr.aria-disabled]="formControl.disabled"
    (focus)="inputFocus($event)"
    (blur)="inputBlur($event)"
  />
  <input
    *ngSwitchCase="'select'"
    type="select"
    #inputRef
    class="input-field typography-body-input"
    [formControl]="formControl"
    [id]="controlName"
    [autocomplete]="autocomplete"
    [placeholder]="placeholder"
    [readonly]="readonly"
    [attr.aria-required]="required"
    [attr.aria-invalid]="formControl.invalid"
    [attr.aria-disabled]="formControl.disabled"
    (focus)="inputFocus($event)"
    (blur)="inputBlur($event)"
  />
  ...
</ng-container>
```

**Comment by GBRBD** [comment]
2024 hello

**Comment by acdphome** [comment]
I don't understand why this error is labeled with "feature", "feature under consideration" and "state:blocked". Seven years ago Dzmitry Shylovich pointed out that it is blocked due to #11716. But that one is completed because they wanted to "implement something along #8785 instead". But #8785 is locked due to inactivity. So how can we make progress on this one?

**Comment by muuvmuuv** [comment]
It wasn't locked due to inactivity, it implements through `hostDirective`. Def. agree that Angular-Team can now continue on this/forms again. Stumbled over this due to select[multiple] does not work properly; return value will always be an array etc.

**Comment by nkarczmarzyk** [comment]
same behaviou for checkbox when using it like this

`<input [type]="'checkbox'" [formControl]="formControl" />`

after changing the value it is returned as string

**Comment by lukeweston1234** [comment]
Yeah this is pretty gross. I just ended up rendering differently with control flow instead.

`
@if(column.type === 'checkbox'){
  <input type="checkbox" [formControlName]="column.field" />
} 
@else {
  <input [type]="column.type" [formControlName]="column.field" />
}
`

**Comment by tortuch** [comment]
Hello, @alxhub when are you going to fix this important issue?

**Comment by muuvmuuv** [comment]
Created a new repro with Angular 18, including Zoneless, esbuild, standalone and OnPush detection. Hopefully this gets us further. Also added my select issue.

> https://stackblitz.com/edit/h3wnfe

**Comment by tortuch** [comment]
> Created a new repro with Angular 18, including Zoneless, esbuild, standalone and OnPush detection. Hopefully this gets us further. Also added my select issue.
> 
> > https://stackblitz.com/edit/h3wnfe

In Native HTML select, multiple property is supported?

**Comment by muuvmuuv** [comment]
What do you mean @tortuch ? Ofc it is https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/multiple#select. My issue is that the behaviour is different with multiple added or removed, set to false, null or true. Especially the output ^^..

<details>
<summary>Example here with vanilla js output</summary>

```html
<form method="get" action="#">
  <select multiple name="dwarfs" id="dwarfs">
    <option value="0">grumpy@woodworkers.com</option>
    <option>happy@woodworkers.com</option>
    <option>sleepy@woodworkers.com</option>
  </select>
</form>

```

```js
const s = document.getElementById("dwarfs")
s.oninput = ({ target }) => {
  console.log("Target value:", target.value)

  for (let o of target.selectedOptions) {
    console.log("Option:", o.value)
  }
}
```

</details>

**Comment by JeanMeche** [comment]
The solution to this will (or already is) signal forms: 

```ts
import { JsonPipe } from '@angular/common';
import { ChangeDetectionStrategy, Component, signal } from '@angular/core';
import { form, FormField } from '@angular/forms/signals';


@Component({
  selector: 'app-root',
  imports: [FormField, JsonPipe],
  template: `
    <input [type]="type()" [formField]="f" /> <button (click)="toggleType()">Toggle Type</button>
    <p>Value: {{ f().value() | json }} - type : {{ typeof f().value() }}</p>
  `,
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class App {
  f = form(signal<string | number>('foobar'));
  type = signal<'text' | 'number'>('text');

  toggleType() {
    this.type.set(this.type() === 'text' ? 'number' : 'text');
  }
}
```

**Comment by chrismclarke** [comment]
I'm not sure if anyone still has this issue, but to the best of my knowledge it can be solved in 2 ways

1. Import `FormsModule` which will include [number_value_accessor](https://github.com/angular/angular/blob/main/packages/forms/src/directives/number_value_accessor.ts)

```ts
import { FormsModule } from '@angular/forms';

@Component({
  imports: [FormsModule ]
})
```

This directive automatically binds to `input[type=number]` and coerces to `number | null` type. I believe this should work with modern signal forms as well as other bindings like formControlName, ngModel etc.

2. Directly bind to the input event and just manually, e.g. if using signals 
```html
<input [value]="myVar()" (input)="myVar.set($event.target.valueAsNumber)" />
```

I'm using new signal forms and to avoid importing entire forms module am using the current custom directive (can be imported into any standalone component):
```ts
import { Directive, ElementRef, forwardRef, inject, Renderer2 } from '@angular/core';
import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';

/**
 * Lightweight directive to use with any formField input
 * to automatically cast `input type="number"` to number type
 *
 * This is an alternative to importing the angular `FormsModule`,
 */
@Directive({
  selector: 'input[type=number][formField]',
  standalone: true,
  host: {
    '(input)': 'handleInput($event)',
    '(blur)': 'onTouched()',
  },
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => AppNumberValueAccessor),
      multi: true,
    },
  ],
})
export class AppNumberValueAccessor implements ControlValueAccessor {
  onChange = (_: any) => {};
  onTouched = () => {};

  private _renderer = inject(Renderer2);
  private _elementRef = inject(ElementRef);

  handleInput(event: Event): void {
    const target = event.target as HTMLInputElement;
    const value = target.valueAsNumber;
    this.onChange(isNaN(value) ? null : value);
  }

  writeValue(value: number | null): void {
    const normalizedValue = value == null ? '' : value;
    this._renderer.setProperty(this._elementRef.nativeElement, 'value', normalizedValue);
  }

  registerOnChange(fn: (value: number | null) => void): void {
    this.onChange = fn;
  }
  registerOnTouched(fn: () => void): void {
    this.onTouched = fn;
  }

  setDisabledState(isDisabled: boolean): void {
    this._renderer.setProperty(this._elementRef.nativeElement, 'disabled', isDisabled);
  }
}
```
